README.md
Nim for awk programmers
=======================

A library of GNU awk functions for nim. Standard awk library functions written in and for nim.

- Program in nim using the familair regex-enabled awk toolset.
- For nim programers, a small set of powerful regex tools from the awk world.
- Convert GNU awk scripts to C (and binary executable) without coding in C by using the nim macro language.

Awk and nim can look very similair. Example awk program that prints the word "text":

```awk
BEGIN{
  str = "This is <a href=\"my text\">here</a>"
  if(match(str, "<a href=\"my text\">", dest)) {
    split(dest[0], arr, "\"")
    if(arr[2] ~ /text/)
      print substr(arr[2], 4, length(arr[2]))
  }
}
```

nim version:

```nim
import awk

var str = "This is <a href=\"my text\">here</a>"
if(match(str, "<a href=\"my text\">", dest) > 0):   
  awk.split(dest, arr, "\"")
  if(arr[1] ~ "text"):
    echo awk.substr(arr[1], 3, len(arr[1]) - 1)  
```

Nim compiles to C source, which compiles to a standalone binary executable using gcc. The nim compile (c) and run (-r) command:

```
nim c -r "test.nim"
text
```

Versions
=======

Most of the nim procs in this package deal with awk's regex functionality. 

Two versions are included: awk.nim uses the "re" module and awknre.nim uses the "nre" module.

The re module is significantly faster and recommended, but awknre.nim is included for backwards compat since the first version of this package used it and there may be some differences in regex options.

Functions
=========

~ and !~
--------
Emulate awk's ~ and !~ commands which can be thought of as a regex-enabled version of contains() in nim. 

```nim
proc `~`*(source, pattern: string): bool 
proc `!~`*(source, pattern: string): bool 
```

Nim does not have an equivilent of awk's // to signify a text is regex. Therefore all text to the right of ~ is treated as regex. To do a literal 
string test use == instead of ~

Use grouping () when building a string with '&', for example:
```nim
if s ~ ("^" & re & "$"):
```

Example:

```nim
import awk
if "george" ~ "ge.*?rge":
  echo "true" #=> true
```

\>* and >>
--------
Write text to a file (append or overwrite)

```nim
proc `>*`(text, filename: string): bool
```
Write `text` to `filename`, overwrite previous content. Close on finish.

```nim
proc `>>`(text, filename: string): bool
```
Append `text` to `filename`. Close on finish.

Example:
```nim
"Hello" & " world" >* "/tmp/test.txt"
"Hello" >* "/dev/stderr"
```

Note that awk's ">" is refactored as ">*" to avoid conflicting with nim's ">" 


match
------
Find regex `pattern` in `source` and optionally store result in `dest`.

```nim
proc match(source: string, pattern: string [, dest: string]): int
```

- `source` is the string to match against.
- `pattern` is the regex pattern.
- `dest` is an optional string to hold the matched text. 
- If `dest` was not declared previously, it will be created. If it exists, match() will overwrite the contents with the results of the match.
- The return value is the number of characters from the start, starting with 1, where the matched text is located, or 0 if no match.
- Consider using index() instead assuming `pattern` is not a regex and not using `dest`, it's faster.

Example:

```nim
import awk
if match("this is a test a", "s.*?a", a) > 0:
  echo a #=> "s is a"   
```

split
--------
Split `source` along regex `match` and store segments in `dest`.

```nim
template split(source: string, dest: untyped, match: string): int
```

- `source` is the source string to be split.
- `dest` is a seq[] filled with results of the split.
- `match` is a string (regex or not) that will be used to split `source`

The function behaves much like awk:

- Returns the number of splits (discardable). 
- The `dest` seq is created by split, it does not need to exist before calling split().
- If the seq does exist, the contents will be overwritten.
- If there are 0 splits `dest` will be 0-length ie. check the return value of split and/or length of `dest` before accessing `dest`
- The first element of `dest` is 0 (unlike awk which is 1). 
- Because nim's system.split() has the same order and type of arguments it should be invoked as awk.split() to avoid ambiguity.

Example:
```nim
import awk
awk.split("This is a string", arr, "is")
echo arr[0] #> "Th"
```

gsub
-----
Global substitute the regex `pattern` with `replacement` in the `source` string

```nim
gsub(pattern: string, replacement: string, source: string): string
```

- `pattern` is a regex string. For literal strings use gsubs()
- `replacement` is the new text to replace the pattern text.
- `source` is the source string.

gsub() returns the new string in addition to changing the source string in-place. It is discardable. 

If the source string is not a var (let, const or literal string) the source string is not modified in-place.

Example 1:
```nim
str = "this is is string"
gsub("[ ]is.*?st", " is a st", str)   
echo str #=> "this is a string"
```

Example 2:
```nim
echo gsub("[ ]is.*?st", " is a st", "this is is string")   
=> "this is a string"
```

Caution: a self-reference will not produce expected results. For example this doesn't produce an error but doesn't work:
```nim
str = "abc"
str = gsub("b", "z", str)
```

gsubi
-----
Global substitute the regex `pattern` with `replacement` in the `source` string, leaving the source string unmodified

```nim
gsubi(pattern: string, replacement: string, source: string): string
```

- `pattern` is a regex string. For literal strings use gsubs()
- `replacement` is the new text to replace the pattern text.
- `source` is the source string.

gsubi() returns the new string but leaves the source string untouched.

Example 1:
```nim
str = "this is is string"
echo gsubi("[ ]is.*?st", " is a st", str)  #=> "this is a string"
echo str #=> "this is is string"
```

gsubs
-----
Global substitute non-regex `pattern` with `replacement` in the `source` string. A literal-string version of gsub()

```nim
gsubs(pattern: string, replacement: string, source: string): string
```

- `pattern` is a literal string
- `replacement` is the new text to replace the pattern text.
- `source` is the source string.

gsubs() returns the new string in addition to changing the source string in-place. It is discardable. 

Example 1:
```nim
str = "this is is string"
gsubs(" is is st", " is a st", str)   
echo str #=> "this is a string"
```

Example 2:
```nim
echo gsubs(" is is st", " is a st", str)   
=> "this is a string"
```

sub
----
```nim
sub(pattern: string, replacement: string, source: string [, occurance: int]): string
```

Substitute in-place the first occurance of regex `pattern` with `replacement` in `source` string
Optional `occurance` substitute at the Xth occurance.

- `pattern` is a regex used in making the substitution
- `replacement` the new string 
- `source` is the string matched against
- `occurance` optional (default 1) which occurance to substitute

If `source` is not a pre-declared variable, sub returns the new string but does not sub in-place
Substitutions are non-overlap eg. sub("22","33","222222") => "333333" not "3333333333"

Example:
```nim
str = "This is a sring"
sub("[ ]is[ ]", " or ", str)                       # substitute 'str' in-place.
echo str #=> "This or a string"
echo sub("[ ]is[ ]", " or ", "This is a sring")    # doesn't sub "This is a sring" in-place, returns a new string
```

subs
-----
Single substitute non-regex `pattern` with `replacement` in the `source` string. A literal-string version of sub(). See gsubs() for documentation


patsplit
--------
Divide `source` into pieces defined by regex `pattern` and store the pieces in seq `field`. Optional `sep` stores the seperators.

```nim
patsplit(source: string, field: seq, pattern: string [, sep: seq]): int
```

- `source` is the source string
- `field` is a sequence containing the field pieces
- `pattern` is a regex (or literal) pattern string
- `sep` is a sequence containing the seperator pieces. Optional.

patsplit() behaves as follows:

- The `field` (and `sep`) sequences must be created beforehand (see example how). 
- Returns number of field elements found.
- If no match found, `field` is set to the value of `source`

Example 1:
```nim
var str = "This is <!--comment1--> a string <!--comment2--> with comments."
var field = newSeq[string](0)
if patsplit(str, field, "<[ ]{0,}[!].*?>") > 0:
  echo field[0] #=> "<!--comment1-->"
  echo field[1] #=> "<!--comment2-->"
```

Example 2:
```nim
var ps = "This is <!--comment--> a string <!--comment2--> with comments."
var field, sep = newSeq[string](0)
patsplit(ps, field, "<[ ]{0,}[!].*?>", sep)
echo sep[1] #=> " a string "
echo unpatsplit(field, sep)
```

unpatsplit
----------
Recombine two sequences created by patsplit()

```nim
unpatsplit(field: seq, sep: seq)
```

Given two seq's created by patsplit, recombine into a single string in alternating sequence ie. field[0] & seq[0] & field[1] & seq[1] etc.

If field has more elements than sep, return ""


substr
------
Return `length`-character long substring of `source` starting at char number `start`

```nim
substr(source: string, start: int [, length: int]): str
```

- The first character is 0 (diff from awk which is 1)
- If `length` not present return the string from `start` to end
- If `start` < 0, treat as 0
- If `start` > length of source, return ""
- If `length` < 1, return ""
- Because nim's system.substr() has the same order and type of arguments this proc should be invoked as awk.substr() to avoid ambiguity.

Example:
```nim
echo awk.substr("Hello World", 3)
#> "lo World"
echo awk.substr("Hello World", 3, 2)
#> "lo"
```

index
------
Return the start location (index) of the first occurance of non-regex `target` in `source`

```nim
index(source: string, target: string): int
```

- First character is 0 (not 1 as in awk)
- If none found or error return -1

Example
```nim
var loc = index("This is string", "is")
echo loc #=> 2
```

Techniques
=============

associative arrays
------------------
Awk uses associative arrays. Nim also supports associative arrays, called "tables". 

For example in awk to uniqe a list of words:

```awk
split("Blue Blue Red Green", arr, " ")        # Whoops, let's get rid of the extra "Blue"

for(i in arr)
  uarr[i] = 1
for(i in uarr)
  print i
```

The equivilent in Nim:
```nim
import strutils, tables

var 
  arr = split("Blue Blue Red Green", " ")     # list of words containing a duplicate
  uarr = initTable[string, int]()             # create empty table (associative array) to hold words

for i in arr:                                 # unique the list
  uarr[i] = 1
for j in uarr.keys:                           # print the list     
  echo j
```

Getting started with nim
========================
- [How I Start](http://howistart.org/posts/nim/1) has good instructions for installing nim. It takes 5 minutes and everything is contained in a single directory.
- [Nim Language](http://nim-lang.org/), official website.
- [GNU awk manual](https://www.gnu.org/software/gawk/manual/gawk.html)


CHANGELOG
2023 Dec 15   - add exception handling for when regex string has syntax errors
                  remove usage of the default "result" variable
                  improvements in the subs() and gsubs() procs
2019 Nov 12   - fix bug in split()
                  fix bug in gsubs()
                  add subs()
2019 May 04   - fix bug in match()
                  fix bug in proc '>>'
                  add gsubi()
                  add gsubs()
2017 April 25 - change from nre to re
                  fix bug in patsplit()
2016 April 27 - first upload


src/awk.nim
import strutils except find
import re

discard """

The MIT License (MIT) 

Copyright (c) 2016-2024 by User:GreenC (at en.wikipedia.org)        

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE."""

# Notes  . re.re() uses {} to ignore whitespace, see http://forum.nim-lang.org/t/213
#        . exception messages are sent by default to stderr - this can be modified via showException()

#
# showException
#
#  Display an exception error. Customize to stdout, logfile, all of these etc.
#
proc showException(msg: string) = 
  try:
    stderr.write(msg)
  except Exception as e:
    discard e.msg              # delima: where to send error msgs when there is an error writing to stderr

#
#  >*
#
# Write 'text\n' to 'filename', overwrite previous content. Close on finish.
#  example:
#    "Hello" & " world" >* "/tmp/test.txt"
#    "Hello" >* "/dev/stderr"
#
proc `>*`*(text, filename: string): bool {.discardable.} =

  var text = text & "\n"
  try:
    writeFile(filename, text)
  except Exception as e:
    showException(e.msg)
    return false
  return true
    
#
# >>
#
# Append 'text\n' to 'filename'. Close on finish
#
proc `>>`*(text, filename: string): bool {.discardable.} =

  var
    fp: File

  if open(fp, filename, fmAppend):
    try:
      writeLine(fp, text)
    except Exception as e:
      showException(e.msg)
      close(fp)
      return false
    finally:
      close(fp)
  else:
    return false
  return true


#
# Match
#
#  match(source [string], pattern [regex])
#  match(source [string], pattern [regex], dest [string])
#
#  Find regex 'pattern' in 'source'.
#  Find regex 'pattern' in 'source' and store result in 'dest'.
#
#  . An optional 'dest' string is created and filled with the results of the match. If already defined it's contents overwritten.
#  . Return number of characters from start of string the match starts, beginning with 1 (not zero).
#  . Return 0 if no match
#  . Consider using index() instead if 'pattern' is not a regex and not using a 'dest'
#
#  Example:
#    import awk
#    if awk.match("this is a test a", "s.*?a", a) > 0:
#      echo a #=> "s is a"
#
#    Notice `a` was not declared no need for "var a: string"
#
proc match*(source: string, pattern: string): int =

    if source.len < 1 and pattern.len < 1:
      return 1
    if source.len < 1 or pattern.len < 1:
      return 0

    var
      j: tuple[first: int, last: int]

    try:
      j = re.findBounds(source, re.re("(?s)" & pattern, {}) )
    except Exception as e:
      showException(e.msg)
      return 0
     
    if j.first > -1:
      return j.first + 1
    else:
      return 0

proc makeDiscardable[T](a: T): T {.discardable.} = a

template match*(source, pattern: string, dest: untyped): int =

    if source.len < 1 and pattern.len < 1:
      makeDiscardable(1)
    if source.len < 1 or pattern.len < 1:
      makeDiscardable(0)

    when compiles(dest):      
      dest = ""
    else:
      var dest = ""

    var
      j: tuple[first: int, last: int]

    try:
      j = re.findBounds(source, re.re("(?s)" & pattern, {}) )
    except Exception as e:
      showException(e.msg)
      makeDiscardable(0)

    if j.first > -1:
      dest = system.substr(source, j.first, j.last)
      makeDiscardable(j.first + 1)
    else:               
      makeDiscardable(0)

#
#   ~ and !~
#
# regex (not) equal to
#
# Example:
#
#  if "george" ~ "ge.*?rge":
#    echo "true" #=> true
#
#  Use () when building a string with '&' eg.
#   if s ~ ("^" & re & "$"):
#
proc `~`*(source, pattern: string): bool =

    if source.len < 1 and pattern.len < 1:
      return true  
    if source.len < 1 or pattern.len < 1:
      return false

    var r = 0

    try:
      r = re.find(source, re.re("(?s)" & pattern, {}))
    except Exception as e:
      showException(e.msg)
      return false

    if r != -1:
      return true
    return false                  

proc `!~`*(source, pattern: string): bool =

    if source.len < 1 and pattern.len < 1:
      return false 
    if source.len < 1 or pattern.len < 1:
      return true

    var r = 0

    try:
      r = re.find(source, re.re("(?s)" & pattern, {}))
    except Exception as e:
      showException(e.msg)
      return true

    if r != -1:
      return false
    return true                

#
# Split
#
#   split(source [string], dest [seq], match [regex])
#
#   Split 'source' along regex 'match' and store segments in 'dest'.
#
#   . Returns the number of elements in 'dest'
#   . 'dest' is a seq[] filled with results of the split.
#   . The 'dest' seq is created by split, it does not need to exist before invoking split.
#   . If seq does exist, its contents will be overwritten.
#   . The first element of 'dest' is 0.
#   . If there are 0 splits, 'dest' will be 0-length
#   . Because nim's system.split() has the same order and type of arguments it should be invoked as awk.split() to avoid ambiguity.
#
template split*(source, dest: untyped, match: string): int =

  when compiles(dest):
    try:
      dest = re.split(source, re.re("(?s)" & match, {}))
    except Exception as e:
      showException(e.msg)
      makeDiscardable(0)
  else:
    var dest: seq[string]
    try:
      dest = re.split(source, re.re("(?s)" & match, {}))
    except Exception as e:
      showException(e.msg)
      makeDiscardable(0)

  if source.len == 0 or dest.len == 0:
    makeDiscardable(0)
  else:
    if dest[0] == source:  # no match
      delete(dest, 0)
      makeDiscardable(0)
    else:
      makeDiscardable(dest.len)

#
# Patsplit
#
#   patsplit(source [string], field [seq], pattern [regex])
#   patsplit(source [string], field [seq], pattern [regex], sep [seq])
#
#   Divide 'source' into pieces defined by regex 'pattern' and store the pieces in seq 'field'
#   Optional 'sep' stores the seperators
#
#    . The 'field' (and 'sep') must be created beforehand (see example) using newSeq[string](0)
#    . Returns number of field elements found.
#    . If no match found, 'field' is set to the value of 'source'
#
#  Example 1
#
#     var str = "This is <!--comment1--> a string <!--comment2--> with comments."
#     var field = newSeq[string](0)
#     if patsplit(str, field, "<[ ]{0,}[!].*?>") > 0:
#       echo field[0] #=> "<!--comment1-->"
#       echo field[1] #=> "<!--comment2-->"
#
#  Example 2
#
#   'field' will store the field content: <!--comment--> and <!--comment2-->
#   'sep' will store the separting pieces: "This is ", " a string ", " with comments."
#
#     var ps = "This is <!--comment--> a string <!--comment2--> with comments."
#     var field, sep = newSeq[string](0)
#     patsplit(ps, field, "<[ ]{0,}[!].*?>", sep)
#     echo sep[1] #=> " a string "
#     echo unpatsplit(field, sep)
#
#
proc patsplit*(source: string, field: var seq[string], pattern: string): int {.discardable.} =

  var        
    source = source 
    i,r = 0
    j: tuple[first: int, last: int]

  field = @[""]

  while len(source) > 0:

    try:
      j = re.findBounds(source, re.re("(?s)" & pattern, {}) )
    except Exception as e:
      showException(e.msg)
      return 0

    if j.first > -1:
      i.inc
      field.insert(system.substr(source, j.first, j.last), i)
      source = system.substr(source, j.last + 1, len(source) - 1)
    else:
      if len(source) > 0:                
        source = ""

  if i > 0:
    field.delete(0)
    r = i
  else:
    r = 0

  return r

proc patsplit*(source: string, field: var seq[string], pattern: string, sep: var seq[string]): int {.discardable.} =

  var
    source = source
    i,r = 0
    j: tuple[first: int, last: int]

  field = @[""]
  sep = @[""]

  while len(source) > 0:

    try:
      j = re.findBounds(source, re.re("(?s)" & pattern, {}) )
    except Exception as e:
      showException(e.msg)
      return 0

    if j.first > -1:
      i.inc
      field.insert(system.substr(source, j.first, j.last), i)
      sep.insert(system.substr(source, 0, j.first - 1), i)
      source = system.substr(source, j.last + 1, len(source) - 1)
    else:
      if len(source) > 0:
        sep.insert(source, i + 1)
        source = ""

  if i > 0:
    field.delete(0)
    sep.delete(0)
    r = i
  else:
    field.insert(source, 1)
    field.delete(0)
    r = 0

  return r

#
# Unpatsplit
#
#  unpatsplit(field [seq], sep [seq])
#
#  Given two seq's created by patsplit, recombine into a single string in alternating sequence ie. field[0] & seq[0] & field[1] & seq[1] etc.
#
#   . If field has more elements than sep, return ""
#   . Typically used to recombine a string separated by patsplit()
#
#
proc unpatsplit*(field, sep: seq[string]): string =

   var o = ""
   if field.len > sep.len:
     return o
   for c in 0..field.len - 1:
     if field[c].len > 0:
       o = o & sep[c] & field[c]
     else:
       o = o & sep[c]
   if sep[sep.len - 1].len > 0 and sep.len != field.len:
     o = o & sep[sep.len - 1]
   return o

#
# Sub
#
#  sub(pattern [regex], replacement [string], source [string])
#  sub(pattern [regex], replacement [string], source [string], occurance [int])
#
#  Substitute in-place the first occurance of regex 'pattern' with 'replacement' in 'source' string
#  Substitute the Xth 'occurance'
#
#   . If 'source' is not a declared variable (eg. literal string), sub() returns the new string but does not sub in-place (see example)
#   . Substitutions are non-overlaping eg. sub("22","33","222222") => "333333" not "3333333333"
#
#  Example:
#    str = "This is a sring"
#    sub("[ ]is[ ]", " or ", str)                       # substitute 'str' in-place.
#    echo str #=> "This or a string"
#    echo sub("[ ]is[ ]", " or ", "This is a sring")    # doesn't sub the literal "This is a sring" in-place, returns the new string
#
#
proc sub*(pattern, replacement: string, source: var string): string {.discardable.} =
  if pattern.len == 0 or source.len == 0:
    return source
  var field, sep = newSeq[string](0)
  if patsplit(source, field, pattern, sep) > 0:
    field[0] = replacement
    source = unpatsplit(field, sep)
  return source

proc sub*(pattern, replacement, source: string): string {.discardable.} =
  if pattern.len == 0 or source.len == 0:
    return source
  var field, sep = newSeq[string](0)
  if patsplit(source, field, pattern, sep) > 0:
    field[0] = replacement
    return unpatsplit(field, sep)
  return source

proc sub*(pattern, replacement: string, source: var string, occurance: int): string {.discardable.} =
  if pattern.len == 0 or source.len == 0:
    return source
  var field, sep = newSeq[string](0)
  var p = patsplit(source, field, pattern, sep)
  if p > 0:
    if occurance <= p:
      field[occurance - 1] = replacement
      source = unpatsplit(field, sep)
  return source

proc sub*(pattern, replacement, source: string, occurance: int): string {.discardable.} =
  if pattern.len == 0 or source.len == 0:
    return source
  var field, sep = newSeq[string](0)
  var p = patsplit(source, field, pattern, sep)
  if p > 0:
    if occurance <= p:
      field[occurance - 1] = replacement
      return unpatsplit(field, sep)
  return source

#
# Gsub
#
#  gsub(pattern [regex], replacement [string], source [string])
#
#  Global substitute in-place the regex 'pattern' with 'replacement' in 'source' string
#
#  Example 1:
#   str = "this is is string"
#   gsub("[ ]is.*?st", " is a st", str)
#   echo str #=> "this is a string"
#
#  Example 2:
#   echo gsub("[ ]is.*?st", " is a st", "this is is string")
#   => "this is a string"
#
#  Caution: a self-reference will not produce expected results eg:  str = gsub(a, b, str) - use gsubi()
#
#
proc gsub*(pattern, replacement: string, source: var string): string {.discardable.} =
  if pattern.len == 0 or source.len == 0:
    return

  try:
    source = re.replace(source, re.re("(?s)" & pattern, {}), replacement)
  except Exception as e:
    showException(e.msg)
    return source

  return source

proc gsub*(pattern, replacement, source: string): string =

  if pattern.len == 0 or source.len == 0:
    return source

  var r = ""
  try:
    r = re.replace(source, re.re("(?s)" & pattern, {}), replacement)
  except Exception as e:
    showException(e.msg)
    return source

  return r

#
# Gsubi
#
#  gsubi(pattern [regex], replacement [string], source [string])
#
#  Global substitute the regex 'pattern' with 'replacement' in 'source' string
#  Returns the new string, leaving the source string unmodified
#
#  Example 1:
#   str = "this is is string"
#   echo gsubi("[ ]is.*?st", " is a st", str)  #=> "this is a string"
#   echo str #=> "this is is string"
#
#
proc gsubi*(pattern, replacement, source: string): string =

  if pattern.len == 0 or source.len == 0:
    return source

  var r = ""
  try:
    r = re.replace(source, re.re("(?s)" & pattern, {}), replacement)
  except Exception as e:
    showException(e.msg)
    return source

  return r

#
# Gsubs
#
# A literal string version of gsub()
#
#  Mimic strutils.replace()
#
proc gsubs*(pattern, replacement: string, source: var string): string {.discardable.} =

    let patLen = pattern.len
    if patLen == 0:
      return source
    if source == "":
      return

    var
      r = ""
      i, j = 0

    while true:
      try:
        # j = boyerMooreHorspool(source, pattern, i)
        j = strutils.find(source, pattern, i)
      except Exception as e:
        showException(e.msg)
        return source
      if j < 0:
        break
      add r, substr(source, i, j - 1)
      add r, replacement
      i = j + patLen
    add r, substr(source, i)
    source = r
    return r

proc gsubs*(pattern, replacement, source: string): string =

    let patLen = pattern.len
    if patLen == 0:
      return source
    if source == "":
      return

    var
      r = ""
      i, j = 0

    while true:
      try:
        # j = boyerMooreHorspool(source, pattern, i)
        j = strutils.find(source, pattern, i)
      except Exception as e:
        showException(e.msg)
        return source
      if j < 0:
        break
      add r, substr(source, i, j - 1)
      add r, replacement
      i = j + patLen
    add r, substr(source, i)
    return r

#
# Subs
#
#   subs(pattern [string], replacement [string], source [string])
#
# Non-regex version of sub(). Replaces first occurance of 'pattern' with 'replacement' in 'source'
#
#   . If 'source' is a literal string (not a variable) returns new string
#     otherwise returns new string (discardable) and modifies 'source' in place
#
# Example:
#   var s = "xxabxx"
#   echo subs("ab", "AB", s) ==> "xxABxx"
#   echo s ==> "xxABxx"
#   echo subs("ab", "AB", "xxabxx") ==> "xxABxx" 
#

proc subs*(pattern, replacement: string, source: var string): string {.discardable.} =

    let patLen = pattern.len
    if patLen == 0:
      return source
    if source == "":
      return

    var
      r = ""
      i, j = 0        

    try:
      # j = boyerMooreHorspool(source, pattern, i)
      j = strutils.find(source, pattern, i)
    except Exception as e:
      showException(e.msg)
      return source 
    if j < 0:
      return source
    add r, substr(source, i, j - 1)
    add r, replacement
    i = j + patLen
    add r, substr(source, i)
    source = r
    return r

proc subs*(pattern, replacement, source: string): string =

    let patLen = pattern.len
    if patLen == 0:
      return source
    if source == "":
      return               

    var
      r = ""
      i, j = 0    

    try:
      # j = boyerMooreHorspool(source, pattern, i)
      j = strutils.find(source, pattern, i)
    except Exception as e:
      showException(e.msg)
      return source
    if j < 0:
      return source
    add r, substr(source, i, j - 1)
    add r, replacement
    i = j + patLen
    add r, substr(source, i)
    return r

#
#
# Substr
#
#   substr(source [string], start [int])
#   substr(source [string], start [int], length[int])
#
# Return 'length'-character long substring of 'source' starting at char number 'start'
#
#   . The first character is 0.
#   . If 'length' not present return the string from 'start' to end
#   . If 'start' < 0, treat as 0
#   . If 'start' > length of source, return ""
#   . If 'length' < 1, return ""
#   . Because nim's system.substr() has the same order and type of arguments this proc should be invoked as awk.substr() to avoid ambiguity
#
# Example:
#   echo awk.substr("Hello World", 3)
#   > "lo World"
#   echo awk.substr("Hello World", 3, 2)
#   > "lo"
#
#
proc substr*(source: string, a: varargs[int]): string =

  var 
    length, start = -1
    alen = a.len

  if alen < 1:
    return ""
  elif alen == 1:
    start = a[0]
    length = source.len
  elif alen == 2:
    start = a[0]
    length = a[1]

  if length > source.len:
    length = source.len
  if start < 0:
    start = 0
  if start > source.len or length < 1 or source.len < 1:
    return ""

  var 
    newsource = source
    final = ""

  for i in 0..source.len - 1:
    if i >= start and i < start + length:
      add(final, newsource[i])
  if len(final) > 0:
    return final
  else:
    return ""

#
# Index
#
#  index(source [string], target [string])
#
# Return 0-oriented start location (index) of first occurance of the non-regex 'target' in 'string'
#
#    . if none found or error return -1
#
# Example
#
#    var loc = index("This is string", "is")
#    echo loc #=> 2
#
#
proc index*(source, target: string): int =

  if source.len == 0 or target.len == 0:
    return -1
  var r = 0
  try:
    r = strutils.find(source, target)
  except Exception as e:
    showException(e.msg)
    return 0
  return r

#
# An implementation of Boyer-Moore-Horspool string searching
#  From Cello https://unicredit.github.io/cello/
#  Copyright (2019) Andrea Ferretti, Apache License 2.0
#  mimic strutils.find() 
#
#  NOT IN USE - reference only. Doesn't work.
#
proc boyerMooreHorspool*(target, query: string, start = 0): int =
  let
    m = len(query)
    n = len(target)
  if m > n: return -1
  var skip = newSeq[int](257)
  for i in 1 .. 256:
    skip[i] = m
  for k in 0 ..< (m - 1):
    skip[query[k].int] = m - k - 1
  var k = start + m - 1
  while k < n:
    var
      j = m - 1
      i = k
    while j >= 0 and target[i] == query[j]:
      dec(j)
      dec(i)
    if j == -1:
      return i + 1
    k += skip[target[k].int]
  return -1
